lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -3364,6 +3364,32 @@ LogicalResult ConvertAtenOp<AtenWhereSelfOp>::matchAndRewrite(
   return success();
 }
 
+template <>
+LogicalResult ConvertAtenOp<AtenCumsumOp>::matchAndRewrite(
+    AtenCumsumOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
+  // Not a tensor type.
+  auto selfType = adaptor.getSelf().getType().dyn_cast<TensorType>();
+  if (!selfType)
+    return rewriter.notifyMatchFailure(
+        op, "Only tensor types input are currently supported");
+  int64_t cumsumDim;
+  if (!matchPattern(op.getDim(), m_TorchConstantInt(&cumsumDim)))
+    return rewriter.notifyMatchFailure(op,
+                                       "non-const dim parameter unsupported");
+
+  if (!op.getDtype().getType().template isa<Torch::NoneType>())
+    return rewriter.notifyMatchFailure(op,
+                                       "Only default dtype none is supported");
+
+  auto outType = getTypeConverter()->convertType(op.getType());
+  rewriter.replaceOpWithNewOp<tosa::ReduceSumOp>(
+      op, outType, adaptor.getSelf(), rewriter.getI64IntegerAttr(cumsumDim));
```

Comment:
This doesn't look right. TOSA ReduceSum is a different operator than cumsum. The TOSA operator is a reduction, so the output tensor along the reduced axis has size 1, where cumsum is still size N, with each value in the summation axis has the partial sum to that point. I don't have an immediate trivial sequence of TOSA operators that can replace cumsum. 

An end to end test with this legalization should fail due to this problem. I'll try to think of a sequence that would work.

Reply:
Yes, this implementation is wrong. So I change this patch to a draft. Still working on this.

Reply:
Someone on my team pointed out that the best legalization is to construct a constant matrix with appropriate weights of 0 or 1, and then you can use TOSA MATMUL. For >2d, you'd have to flatten the original tensor, but if you only need to handle 1 or 2d tensors, it's not that bad.

Reply:
I am planning to so tosa::Slice(start=0, step=1:dimShape) + tosa::ReduceSum(sliced tensor, dim) + tosa::Concat(ReduceSum tensors) to solve it once for all.

Reply:
That should work. If you are calculating cumsum across a big tensor, you will have a big instruction stream, but the solution with MATMUL would have a big constant tensor, so they may not be that different.

Reply:
TOSA should really have a cumsum operator though since this decomposition is an algorithmic lowering decision that we should be baking in here at the frontend level.

If we cannot add that op, we should do it with the algorithm shown below with log2 reduction steps:
Run with `DEBUG = True` for more intuition for the algorithm.


```
import numpy as np

DEBUG = False

def cumsum_log2(x):
    """Compute cumsum with log2 reduction steps."""
    orig_size = x.shape[0]
    log2 = int(np.ceil(np.log2(orig_size)))
    padded_size = 2**log2
    x_padded = np.pad(x, (0, padded_size - orig_size), constant_values=0)
    cumsum = np.copy(x_padded)
    if DEBUG:
        print(f"input: {x_padded}")
    for i in range(log2):
        summand = np.reshape(np.copy(x_padded), (2**i, 2, 2**(log2 - i - 1)))
        summand[:, 1, :] = np.sum(summand[:, 0, :], axis=1, keepdims=True)
        summand[:, 0, :] = 0
        if DEBUG:
            print(f"summing: {summand.flatten()}")
        cumsum += summand.flatten()
    if DEBUG:
        print(f"result: {cumsum}")
    return cumsum[:orig_size]


def check(x):
    ours = cumsum_log2(x)
    reference = np.cumsum(x)
    if not np.allclose(ours, reference):
        print(f"FAILED: {x}")
        print(f"OURS:      {ours}")
        print(f"REFERENCE: {reference}")
    else:
        print("PASSED")

check(np.ones(7))
check(1. + np.linspace(0.0, 10.0, 6))
check(np.random.randn(600))
```

Reply:
```
log2 = int(np.ceil(np.log2(orig_size))) # tosa.ceil (mlir::tosa::CeilOp)  tosa.log (mlir::tosa::LogOp) 
padded_size = 2**log2 # tosa.exp (mlir::tosa::ExpOp) 
x_padded = np.pad(x, (0, padded_size - orig_size), constant_values=0) # tosa.pad (mlir::tosa::PadOp)
summand = np.reshape(np.copy(x_padded), (2**i, 2, 2**(log2 - i - 1))) # tosa.reshape (mlir::tosa::ReshapeOp) 
summand[:, 1, :] = np.sum(summand[:, 0, :], axis=1, keepdims=True) # tosa.slice (mlir::tosa::SliceOp)
```
@silvasean  Here is the corresponding tosa ops that I can find and translate from your python code.
But the log2 is the return mlir::Value of tosa.ceil, then if I want to use it in the `for i in range(log2)`, I will have to get the exact int value of log2, so I can write a for loop in c++. **How can I get this int value from mlir::value?** 
In previous code, whenever I need to get the int  value of mlir::Value, I use something like this: 
```
  int64_t int_min;
  if (!matchPattern(op.getMin(), m_TorchConstantInt(&int_min)))
    return rewriter.notifyMatchFailure(
        op, "unimplemented: value `int_min` should be a torch constant int");
```
But it is not the same case in this issue. Should I try something like ` int log2 =  tosaCeilReturnValue.getValue().getSExtValue();`?
Otherwise, should we manually implement the c++ ceil and log helper function so we can get rid of using tosa::log and tosa::ceil?

Reply:
I'll look at adding CUMSUM to TOSA. It isn't ruled out, although it does trend against the principle of "If the operator can be broken down, then we should look at the component operators." It's not an absolute rule, so it's a matter of defining the rationale for the exception.

Reply:
@AmosLewis you can manually unroll the computation in the compiler since it is only log2 steps. So you would calculate log2 inside the compiler in C++ based on the static shape (since TOSA only supports static shapes anyways).

@eric-k256 There is a strong argument here for providing the CUMSUM op, since there are many different ways to lower it. In particular, the logarithmic reduction I showed here is known to be suboptimal for many hardware, since the smaller reduction steps will likely want to be done with the equivalent of a warp-level cumsum operation, which is very difficult to reconstruct from the code we lower here. Other approaches could be to do some number of logarithmic reduction steps and then a batch matmul for the smaller ones. And if there is a bunch of parallelism in the non-summed dimensions then a more straightforward serialized cumsum algorithm would be very efficient (which is not efficiently representable in TOSA for current compilers, since it would require turning a TOSA WHILE operator into a within-dispatch loop rather than a host-side loop).

Reply:
@eric-k256  How long do you think it will take for adding tosa::cumsum? I am currently waiting for this op. Before the tosa::cumsum merged, I am planning to use my concat(reducesum(slice)) implementation locally and continue on fix the following up other ops(broadcast_to/....) in the OPT model.